frame/elections-phragmen/src/migrations/unlock_and_unreserve_all_funds.rs

```diff
@@ -0,0 +1,494 @@
+			.iter()
+			.map(|account| {
+				(
+					account.clone(),
+					Self::get_actual_locked_amount(account.clone()).unwrap_or(Zero::zero()),
+				)
+			})
+			.collect();
+
+		let account_reserved_before: BTreeMap<T::AccountId, BalanceOf<T>> = all_accounts
+			.iter()
+			.map(|account| (account.clone(), T::Currency::reserved_balance(&account)))
+			.collect();
+
+		// Total deposited for each account *should* be less than or equal to the total reserved,
+		// however this does not hold for all cases due to bugs in the reserve logic of this pallet.
+		let bugged_deposits = all_accounts
+			.iter()
+			.filter(|account| {
+				account_deposited_sums.get(&account).unwrap_or(&Zero::zero()) >
+					account_reserved_before.get(&account).unwrap_or(&Zero::zero())
+			})
+			.count();
+
+		// The calculation of the locked & staked logic is not known to be bugged, so we assert
+		// that the amounts staked in the pallet match the amounts actually locked in Balances.
+		// If this fails, there is likely a bug in the migration and it should be investigated.
+		ensure!(
+			!all_accounts
+				.iter()
+				.any(|account| account_locked_before.get(&account).unwrap_or(&Zero::zero()) !=
+					account_staked_sums.get(&account).unwrap_or(&Zero::zero())),
+			"account locked != account staked"
+		);
+
+		// Print some summary stats.
+		let total_stake_to_unlock = account_staked_sums.clone().into_values().sum::<BalanceOf<T>>();
+		let total_deposits_to_unreserve =
+			account_deposited_sums.clone().into_values().sum::<BalanceOf<T>>();
+		log::info!("Total accounts: {:?}", all_accounts.len());
+		log::info!("Total stake to unlock: {:?}", total_stake_to_unlock);
+		log::info!("Total deposit to unreserve: {:?}", total_deposits_to_unreserve);
+		log::info!("Bugged deposits: {}/{}", bugged_deposits, all_accounts.len());
+
+		let pre_migration_data =
+			PreMigrationData::<T> { account_locked_before, account_reserved_before };
+		Ok(pre_migration_data.encode())
+	}
+
+	/// Executes the migration.
+	///
+	/// Steps:
+	/// 1. Retrieves the deposit and stake amounts from the pallet.
+	/// 2. Unreserves the deposited funds for each account.
+	/// 3. Unlocks the staked funds for each account.
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		// Get staked and deposited balances as reported by this pallet.
+		let (account_deposited_sums, account_staked_sums, initial_reads_weight) =
+			Self::get_account_deposited_and_staked_sums();
+
+		// Deposited funds need to be unreserved.
+		for (account, unreserve_amount) in account_deposited_sums {
+			if unreserve_amount.is_zero() {
+				continue
+			}
+			T::Currency::unreserve(&account, unreserve_amount);
+		}
+
+		// Staked funds need to be unlocked.
+		for (account, amount) in account_staked_sums {
+			if amount.is_zero() {
+				continue
+			}
+			// Question for reviewers:
+			// remove_lock triggers a few
+			// `ERROR main runtime::system: Logic error: Unexpected underflow in reducing consumer`
```

Comment:
This may be related: https://github.com/paritytech/substrate/issues/13114#issuecomment-1377697085

Maybe if you can print the account IDs that are causing the error we can check their state in the staking ledger and confirm @kianenigma's assumption.

Reply:
How do I print the account ids? They log an empty string when I try `log::info!("acc: {:?}", account_id)`.

Reply:
Ah yes, that's because of [`Display` implementation](https://paritytech.github.io/substrate/master/src/sp_core/crypto.rs.html#594) of the AccountId in no-std. If you pass the `--execution native` flag to try-runtime you should be able to see the account IDs.

Reply:
Ah wow I had no idea that flag existed! That'll come in very handy, thanks! I'm OOO tomorrow but will check those addresses on Friday.

Reply:
I wonder if `try-runtime-cli` should run in native mode by default @kianenigma for better debugging? Unless there's some drawback to using it.

Reply:
@liamaharon I may be inducing you in error here as the `--execution native` option has actually been deprecated (it is still part of the [try-runtime substrate docs](https://docs.substrate.io/reference/command-line-tools/try-runtime/), which I keep forgetting it's massively outdated, we should probably just kill that page or redirect to the rust docs page).